Added dynamic routing and redesigned API

mkosieradzki opened the pull request:
Added dynamic routing and redesigned API
Fixed cookie support

This PR addresses:
#37, #35, #34 ?, #27, #10, #43, #47 

NOTES:
- There are no tests for new use-cases I will create them if you accept the new API.
- I will need to add one more callback (probably not to options but to one of the .ForwardRequest overloads to allow modifying response headers before sending back content).

If you are not convinced about merging QueryString parameters - this is an extremely useful feature when forwarding to the Service Fabric Reverse Proxy. You are often adding query string parameters like PartitionKey=.

dnfclas commented:
@mkosieradzki,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

mkosieradzki commented:
Redesign in-progress.

mkosieradzki commented:
@Tratcher , @davidfowl , @mikaelm12 
I would like to know what you think about this guys. If you are OK. I will do final improvements and prepare this for merge.

Final api looks like this:
```
public void ConfigureServices(IServiceCollection services)
{
    services.AddProxy();
}

public void Configure(IApplicationBuilder app)
{
    app.UseWebSockets().RunProxy(new Uri("https://example.com"));
}
```

More advanced APIs
1. Setting options
```
services.AddProxy(options =>
{
    options.MessageHandler = .....
    options.WebSocket....
}
```
2. Dynamic routing
```
app.Run(async ctx =>
{
    /// Uri can be computed based on ctx
    ctx.ProxyRequest(new Uri("https://example.com"));
})
```
3. Total-control (advanced scenarios):
```
            if (context.WebSockets.IsWebSocketRequest)
            {
                var wsDestinationUri = new Uri(ProxyUtils.TranslateUriToWebSocketScheme(destinationUri.ToString()));
                await ProxyUtils.ForwardWebSocketRequest(context, wsDestinationUri);
            }
            else
            {
                using (var requestMessage = ProxyUtils.PrepareHttpRequestMessage(context.Request, destinationUri))
                using (var responseMessage = await ProxyUtils.ForwardHttpRequestMessageToServer(context, requestMessage))
                {
                    await ProxyUtils.ForwardHttpResponseMessageToClient(responseMessage, context.Response);
                }
            }
```
ProxyUtils expose basic reasonable building blocks.

I initally intended to go callback way, but this callback started to look way to similary to app.Run :) I have refactored everything to this.

mkosieradzki commented:
I am considering two improvements:
1. Moving ForwardWebSocketRequest, PrepareHttpRequestMessage, ForwardHttpRequestMessageToServer, ForwardHttpResponseMessageToClient to a new class: AdvancedProxyExtensions, making them extension methods for HttpContext and placing it inside Microsoft.AspNetCore.Proxy namespace.

Also it would be a good idea to rename those methods because they are now in different context:
- ForwardWebSocketRequest => AcceptProxyWebSocketRequest
- PrepareHttpRequestMessage => CreateProxyHttpRequest
- ForwardHttpRequestMessageToServer => SendProxyHttpRequest
- ForwardHttpResponseMessageToClient => ReceiveProxyHttpResponse


Other basic extensions are in Microsoft.Extensions.DependencyInjection and Microsoft.AspNetCore.Builder namespaces and they practically never require Microsoft.AspNetCore.Proxy for basic scenarios.

2. Make TranslateUriToWebSocketScheme operate on Uris instead of strings and move it to AdvancedProxyExtensions maybe make it extension of Uri and rename it to ToWebSocketScheme()

The resulting advanced mode code code will look as below:
```
using Microsoft.AspNetCore.Proxy;

[...]

if (context.WebSockets.IsWebSocketRequest)
{
    await context.AcceptProxyWebSocketRequest(wsDestinationUri.ToWebSocketScheme());
}
else
{
    using (var requestMessage = context.CreateProxyHttpRequest(destinationUri))
    using (var responseMessage = await context.SendProxyHttpRequest(requestMessage))
    {
        await context.ReceiveProxyHttpResponse(responseMessage);
    }
}
```

mkosieradzki commented:
Above are fixes for #47 (Conditionally added WinHttpHandler for .NET 4.6)

Tratcher commented on src/Microsoft.AspNetCore.Proxy/ProxyServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,26 @@
+    {
+        public static IServiceCollection AddProxy(this IServiceCollection services)
+        {
+            return services.AddSingleton(new ProxyService(new SharedProxyOptions()));
```
`AddSingleton<ProxyService>()` and then ProxyService takes an `IOptions<SharedProxyOptions>`

Tratcher replied:
Moving this to a central service and options prevents an app from proxying to more than one back-end that has different settings (e.g. configuring the HttpClientHandler differently). E.g. using different credentials, etc..

mkosieradzki replied:
> Moving this to a central service and options prevents an app from proxying to more than one back-end that has different settings (e.g. configuring the HttpClientHandler differently). E.g. using different credentials, etc..
1. Credentials should be configured at HttpClient not at the handler.
2. But if you really need dedicated handler you can use AddScoped :). So no trade-off here at all

mkosieradzki replied:
> AddSingleton<ProxyService>() and then ProxyService takes an IOptions<SharedProxyOptions> - 

If you really think it is a good idea we can go this way. My concern is that auto-injection is slightly slower, but I can adjust.

Tratcher replied:
Scoping does not help here, that would be per request, not per back end. Also, creating a new HttpClient on each request would cause major perf problems and port exhastion.

mkosieradzki replied:
Good point.
It is workaroundable by providing singleton instances (for example static ones) when configuring service on per-scope-basis, however I think a lot of people can accidentaly misuse it. What can lead to a disaster which you have pointed out.

However, can you think of any legit scenario for having multiple HttpHandlers?
1. If you need conditional logic you always have DelegatingHandler.
2, WinHttpHandler vs ClientHttpHandler is usually a per-app choice. Legacy version of the ClientHttpHandler (4.0.0 or 4.3.1 on .NET 4.6) is not an option in proxy scenario... because #47 

I really cannot think about any scenario (what doesn't mean there is no scenario).

Tratcher commented on src/Microsoft.AspNetCore.Proxy/SharedProxyOptions.cs:
```diff
@@ -0,0 +1,31 @@
+    /// <summary>
+    /// Shared Proxy Options
+    /// </summary>
+    public class SharedProxyOptions
```
Why not just ProxyOptions?

mkosieradzki replied:
> Why not just ProxyOptions?

I have taken this naming from AuthenticationMiddleware. Can go with ProxyOptions as well. What's your final verdict?

mikaelm12 replied:
ProxyOptions seems like a better choice to me

mkosieradzki replied:
OK. ProxyOptions it is :).

Tratcher commented on src/Microsoft.AspNetCore.Proxy/ProxyAdvancedExtensions.cs:
```diff
@@ -0,0 +1,192 @@
+        private static readonly string[] NotForwardedWebSocketHeaders = new[] { "Connection", "Host", "Upgrade", "Sec-WebSocket-Key", "Sec-WebSocket-Version" };
+        private const int DefaultWebSocketBufferSize = 4096;
+
+        public static Uri ToWebSocketScheme(this Uri uri)
```
internal, it's just infrastructure.

mkosieradzki replied:
Please, no. I really, really need those methods (maybe except the ToWebSocketScheme for advanced scenarios addressing for example #35.

They are not visible unless you add use Microsoft.AspNetCore.Proxy; All other classes are in dependency injection related namespaces. And those are intended for advanced use scanarios as I had described. I really put a big effort to design those methods naming in a usable way.

mkosieradzki replied:
Below is the code designed for advanced scenarios. I wanted to use callbacks as in #10 . But the more I was working with different scenarios the more those 4 helper methods seemed to be an optimal advanced abstraction.
```
using Microsoft.AspNetCore.Proxy;

[...]

if (context.WebSockets.IsWebSocketRequest)
{
    await context.AcceptProxyWebSocketRequest(wsDestinationUri.ToWebSocketScheme());
}
else
{
    using (var requestMessage = context.CreateProxyHttpRequest(destinationUri))
    using (var responseMessage = await context.SendProxyHttpRequest(requestMessage))
    {
        await context.ReceiveProxyHttpResponse(responseMessage);
    }
}
```

mkosieradzki replied:
For example using this abstraction you can even implement some retry patterns for idempotent HTTP verbs. I am really open to this discussion, but this API surface decisions comes from practical fork of this repo, which I used for the Service Fabric stateless layer.

Tratcher replied:
I was only referring to the ToWebSocketScheme method

mkosieradzki replied:
Rotfl ;) - total misunderstanding. Funny part is that I have ended up inheriting DelegatingHandler and the code seems much cleaner.

Now I would be the one opting for internalizing entire ProxyAdvancedExtensions... as cannot think of any scenario that couldn't be nicely handled by the DelegatingHandler.

mkosieradzki commented:
@Tratcher , @mikaelm12 Thank you very much for code reviews. I did all the suggested changes, except of making AdvancedProxyExtensions internal. I have explained my point in the response to your review.

If you will approve to leave the AdvancedProxyExtensions public, I will write all the necessary xmldocs.

mkosieradzki commented:
@Tratcher I think you might be right about internalizing this and doing advanced scenarios through the HttpMessageHandler, espacially Delegating handler seems to be a good fit here, I will check it in my test scenarios and will be back soon.

mkosieradzki commented:
@Tratcher It works perfectly well.

From practical point of view the only dowside of the DelegatingHandler is that you have no access to the original HttpContext. But we can pass it through the request.Properties (not obvious enough) or simply we can add the callback mentioned in #35 . But as Func<HttpRequest, HttpRequestMessage, Task>.

Tratcher commented on src/Microsoft.AspNetCore.Proxy/ProxyAdvancedExtensions.cs:
```diff
@@ -0,0 +1,192 @@
+
+namespace Microsoft.AspNetCore.Proxy
+{
+    static class ProxyAdvancedExtentions
```
internal? always be explicit about visibility.

mkosieradzki replied:
OK. Didn't know that convention. Will fix. Sorry for that. If we are talking about being explicit, why are you never explicit about class being "sealed"? That looks really strange but I see this very often.

Tratcher replied:
sealed never the implicit state, a class is virtual by default (though I don't think you can actually mark a class as virtual). In general we do not seal classes as it prevents useful extensibility.

mkosieradzki replied:
In my opinion it is a bad practice not to seal classes. Only very small subset of classes is designed to be inherited. Many modern objective-functional languages even prohibit inheritance, but that's different discussion. OK, nvm. I will comply.

Tratcher commented on src/Microsoft.AspNetCore.Proxy/ProxyAdvancedExtensions.cs:
```diff
@@ -0,0 +1,192 @@
+            return proxyService.Client.SendAsync(requestMessage, HttpCompletionOption.ResponseHeadersRead, context.RequestAborted);
+        }
+
+        public static Task ReceiveProxyHttpResponse(this HttpContext context, HttpResponseMessage responseMessage)
```
CopyProxyHttpResponse

mkosieradzki replied:
Done

Tratcher commented on src/Microsoft.AspNetCore.Proxy/ProxyExtensions.cs:
```diff
@@ -0,0 +1,64 @@
+
+                using (var requestMessage = context.CreateProxyHttpRequest(destinationUri))
+                {
+                    var prepareRequestHandler = proxyService.Options.PrepareRequest;
```
What is this used for? You don't have any tests or samples for it.

mkosieradzki replied:
This is a change introduced for #35 and #4 
Basically in my opinion for most of scenarios (including retries) DelegatingHandler should be used.
I have introduced this PrepareRequest handler in the last commit to allow to altering this request while having access to the original HttpRequest. I am not sure about this one. If you approve it I will add tests and examples.

If you reject it I am also OK with your decision, as for my scenarios DelegatingHandler is more than enough. And in extreme cases AsyncLocal can be used to pass HttpContext to the handler.

So what's your verdict here? I am OK with both.

Regarding #4 I think we should built this in. But I will do this in a separate pull request it deserves a separate discussion. It's quite complex as there is an emerging HTTP specification called Forwarded, and competitive a competitive one defacto standard X-Forwarded-Host... You can learn more here #Azure/service-fabric-issues/issues/113 .

Tratcher replied:
Makes sense, add it to the sample and tests.

Tratcher commented on src/Microsoft.AspNetCore.Proxy/ProxyAdvancedExtensions.cs:
```diff
@@ -0,0 +1,192 @@
+
+            // SendAsync removes chunking from the response. This removes the header so it doesn't expect a chunked response.
+            response.Headers.Remove("transfer-encoding");
+            return responseMessage.Content.CopyToAsync(response.Body);
```
pass HttpContext.RequestAborted as the cancellation token here. That will terminate upstream reads of the client disconnects.

mkosieradzki replied:
HttpContent.CopytoAsync does not have overload with CancellationToken

Tratcher replied:
oh, right. ReadAsStreamAsync and then CopyToAsync then.

davidfowl replied:
We should file a bug on `HttpClient` to add that overload

mkosieradzki replied:
@Tratcher @davidfowl It was code coming from the original code I did only move it, not change it. Can we handle this in a separate issue and connect it the the new issue mentioned by @davidfowl? If your verdict is to do ReadAsStreamAsync  and then CopyToAsync I will fix it immediately. What's your verdict?

Tratcher replied:
Use ReadAsStreamAsync. He was suggesting a separate change to HttpContent itself.

mkosieradzki commented:
> One design change you've made that I don't see discussed is that you now drop the PathBase, Path, and Query when proxying a request. You always proxy to a static address. Or did I miss something?

Fail. I actually accidentaly drop all of this when started oversimplifying the API.

That's perfectly OK for ProxyRequest, but I have broken RunProxy - I will fix it immediately!

Thanks a lot!

mkosieradzki commented:
I will need to revive the ProxyMiddleware for this and the ProxyOptions. But meanwhile current ProxyOptions will become SharedProxyOptions to avoid naming clash.

mkosieradzki commented:
@Tratcher I have fixed those issues, and added some xmldoc. Thanks a lot for pointing this out! I still need your decision regarding PrepareRequest. I don't know whether I should remove it or add tests for it.

Tratcher commented on src/Microsoft.AspNetCore.Proxy/ProxyOptions.cs:
```diff
@@ -1,18 +1,30 @@
+    /// Proxy Options
     /// </summary>
-    public class ProxyOptions
+    public sealed class ProxyOptions
```
No, we do not use sealed.

mkosieradzki commented:
@Tratcher Done. I have added XFH (which is quite useful) as an example, which was suggested by #4

Tratcher commented on src/Microsoft.AspNetCore.Proxy/ProxyService.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace Microsoft.AspNetCore.Proxy
+{
+    internal class ProxyService
```
public

Tratcher commented on src/Microsoft.AspNetCore.Proxy/SharedProxyOptions.cs:
```diff
@@ -0,0 +1,46 @@
+        /// Allows to modify HttpRequestMessage before it is sent to the Message Handler.
+        /// </summary>
+        public Func<HttpRequest, HttpRequestMessage, Task> PrepareRequest { get; set; }
+        /// <summary>
```
nit: new line before comments

mkosieradzki commented:
I have merged latest changes from #41 and temporarily removed workarounds for #47 as they were breaking IIS compatibility hard way (ceased sample project to work). I will create a separate pull request for that issue later.

mkosieradzki commented:
@mikaelm12 I have merged upstream branch. Ready to merge.

mikaelm12 commented on src/Microsoft.AspNetCore.Proxy/ProxyServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+﻿
```
Need a copyright and license header here

mkosieradzki commented:
@mikaelm12 Thanks, any idea why is Travis failing? Current source code is failing as well. This specific PR is not introducing any new failed test. 

Shouldn't we fix travis in a separate PR, because this one is huge enough ;).

It looks like a problem with ClientWebSocket on non-windows platform. I was trying to check other WebSocket-related repos but they have even AppVeyor failing.

mikaelm12 commented:
Any ideas @davidfowl @anurse?
We're getting an exception in System.Net.Websockets saying that its unable to connect 
https://travis-ci.org/aspnet/Proxy/jobs/209152222#L1859

mkosieradzki commented:
@mikaelm12 Just letting you know this is ready to merge.

mikaelm12 commented:
@mkosieradzki I haven't forgotten about this 😄 
I need to investigate the flaky WebSocket test before merging this.

mkosieradzki commented:
@mikaelm12 But this one has nothing to do with the WebSocket test, WebSockets were in the previous commit (but the test didn't fail because travis was down).

However I can try to work on the websockets issue.

KasperHoldum commented:
Any progress on dynamic routing?

mkosieradzki commented:
I have updated this PR and fixed one more issue (cookie support - 1 liner).

@mikaelm12  - please merge if possible (again this PR has nothing to do with the flaky WebSockets test).

BTW. I can try to investigate this flaky test (but it is unrelated to this PR).

davidfowl commented:
Why is there a public ProxyService class?

mkosieradzki commented:
@davidfowl 
I honestly don't know. I did it becase @Tratcher explictly asked me to ;). My wild guess is to be able to add remove service manually from DI container. It is registered directly under its type (not interface).  So if it was internal there will be no means to "address" this service. But that's only my guess.

davidfowl commented:
@Tratcher  can you clarify?

@mkosieradzki It doesn't look as though it's required at all. Also, you didn't rename SharedProxyOptions to ProxyOptions

mkosieradzki commented:
@davidfowl Please see (Feb 28):
> I will need to revive the ProxyMiddleware for this and the ProxyOptions. But meanwhile current ProxyOptions will become SharedProxyOptions to avoid naming clash.

Tratcher commented:
I don't recall the reasoning for using a service. The only discussion I can find for it is here: https://github.com/aspnet/Proxy/pull/46#discussion_r103326303

You should be able to remove the service, inject the options into the middleware, and flow them from there.
